metric_config_parser/config.py

```diff
@@ -426,6 +359,146 @@ def from_github_repos(
+                repo.iter_commits("HEAD", paths=default_config_file)
+            ).committed_date
+
+            default_config_json = toml.load(default_config_file)
+
+            if "project" in default_config_json:
+                # opmon spec
+                spec = MonitoringSpec.from_dict(default_config_json)
+            else:
+                spec = AnalysisSpec.from_dict(default_config_json)
+                spec.experiment.is_private = spec.experiment.is_private or is_private
+
+            default_configs.append(
+                DefaultConfig(
+                    default_config_file.stem,
+                    spec,
+                    UTC.localize(dt.datetime.utcfromtimestamp(last_modified)),
+                    is_private=is_private,
+                )
+            )
+
+        definitions = []
+        for definitions_config_file in files_path.glob(f"{DEFINITIONS_DIR}/*.toml"):
+            last_modified = next(
+                repo.iter_commits("HEAD", paths=definitions_config_file)
+            ).committed_date
+
+            definitions.append(
+                DefinitionConfig(
+                    definitions_config_file.stem,
+                    DefinitionSpec.from_dict(toml.load(definitions_config_file)),
+                    UTC.localize(dt.datetime.utcfromtimestamp(last_modified)),
+                    platform=definitions_config_file.stem,
+                    is_private=is_private,
+                )
+            )
+
+        functions_spec = None
+        for functions_file in files_path.glob(f"{DEFINITIONS_DIR}/{FUNCTIONS_FILE}"):
+            functions_spec = FunctionsSpec.from_dict(toml.load(functions_file))
+
+        return cls(
+            external_configs,
+            outcomes,
+            default_configs,
+            definitions,
+            functions_spec,
+            repos=[Repository(repo=repo, path=path, main_branch=main_branch)],
+            is_private=is_private,
+        )
+
+    def as_of(self, timestamp: datetime) -> "ConfigCollection":
+        """Get configs as they were at the provided timestamp."""
+        config_collection = None
+
+        # configs can be loaded from multiple different repos
+        for repo in self.repos:
+            # copy the original repo to a temporary path were we can go back in history
+            with TemporaryDirectory() as tmp_dir:
+                git_dir = Path(repo.repo.git_dir)
+                shutil.copytree(git_dir.parent, tmp_dir, dirs_exist_ok=True)
+                tmp_repo = Repo(tmp_dir)
+
+                # find the commit that got added just before the `timestamp`
+                rev = "HEAD"
+                tmp_repo.git.checkout(repo.main_branch)
+                for commit in tmp_repo.iter_commits("HEAD"):
+                    if commit.committed_datetime <= timestamp:
+                        rev = commit.hexsha
+                        break
+
+                if commit:
+                    rev = commit.hexsha
```

Comment:
This block is confusing me. Is it just for the case where we found a commit that is older than HEAD but didn't find one older than `timestamp`? So in that case we use the closest we can find instead of just defaulting to HEAD?

Reply:
I added some more comments to the code. But yes, if there is no commit that is older than the reference timestamp, use the oldest commit available